inputremapper/injection/macros/macro.py

```diff
@@ -540,12 +541,18 @@ async def task(handler: Callable):
 
         self.tasks.append(task)
 
-    def add_wait(self, time: Union[int, float]):
+    def add_wait(self, time: Union[str, int], vary: int=0):
         """Wait time in milliseconds."""
-        time = _type_check(time, [int, float], "wait", 1)
+        time = _type_check(time, [int], "wait", 1)
+        vary = _type_check(vary, [int], "wait", 2)
 
         async def task(_):
-            await asyncio.sleep(_resolve(time, [int, float]) / 1000)
+            if _resolve(vary, [int]) > _resolve(time, [int]):
+                time_vary = random.randint(_resolve(time, [int]), _resolve(vary, [int]))
+            else:
+                time_vary = _resolve(time, [int])
+
+            await asyncio.sleep(_resolve(time_vary, [int, float]) / 1000)
```

Comment:
`_resolve` makes rpc calls. Though I doubt it would be noticeable by a user, on paper it would most likely be a lot better for performance if `_resolve` was called once for `vary` and `time`, and then the result stored in a local variable within `task` to reference it later on, instead of resolving `time` redundantly three times. It also reduces the risk of race conditions.

Reply:
Fair. When I made the initial changes the `vary > time` comparison didn't work and my knowledge of python isn't spectacular.

Reply:
you need to resolve in in `task`. `_resolve` looks up variables, and they can change during runtime. If you do it outside of `_task` they are resolved when the macro is parsed, and then never again.

Reply:
I've moved the `_resolve` into `_task`. I'm not sure how to set the default value for `max_time` to be `None` as it throws an error comparing `'NoneType' and 'int'` when the second argument isn't set. Again, my python is not amazing.

A slightly off-topic aside: In testing I found a way to test if macro variable is set or not by `if_eq($var, None, then=unset_macro, else=set_macro)`. Would it be worth amending the examples documentation to illustrate it as a roundabout `if_unset()` function? Using `wait` with an unset variable can softlock keys if you have something like `key_down(a).wait($unsetvariable).key_up(a)` as the `wait` dies so `key_up(a)` never fires and having the macro always set `unsetvariable` first negates changing the value elsewhere.

Reply:
> as it throws an error comparing 'NoneType' and 'int' when the second argument isn't set. 

yeah, you would need to do something like `if max_time is not None and max_time > ...`

this way, you are not comparing `'NoneType' and 'int'`, because it never reaches the comparison if max_time is None.

---

Do you mean something like

```
if_eq(
  $unsetvariable,
  None,
  then=?,
  else=set(unsetvariable, 1000)
).key_down(a).wait($unsetvariable).key_up(a)
```

?

Reply:
So `_resolve` should get a `default` parameter? Yeah, maybe. But that should be done in a separate pull request, and I'd want to have another opinion on that.

for now, we still need a new unit test. Take a look at `test_0`, and modify it to add a wait to the front. I'd like to see three test-cases:
- `wait(50).key(a)`
- `wait(50, 100).key(a)`
- `set(a, 100).wait(50, $a).key(a)`

because I think there might be a bug in there.

You could time the time it takes for maybe 5 runs of `await macro.run(self.handler)` to run, and from that figure out if the average time it was 50, or some value between 50 and 100.

Reply:
You can run macro tests via

```
sudo pkill -f input-remapper-service
python3 tests/test.py tests.unit.test_macros
```

Reply:
Adding

```
    async def test_wait_1(self):
        macro = parse("wait(10).key(1)", self.context, DummyMapping, True)
        one_code = system_mapping.get("1")

        await macro.run(self.handler)
        self.assertListEqual(
            self.result,
            [(EV_KEY, one_code, 1), (EV_KEY, one_code, 0)],
        )
        self.assertEqual(len(macro.child_macros), 0)

    async def test_wait_2(self):
        macro = parse("wait(10,100).key(1)", self.context, DummyMapping, True)
        one_code = system_mapping.get("1")

        await macro.run(self.handler)
        self.assertListEqual(
            self.result,
            [(EV_KEY, one_code, 1), (EV_KEY, one_code, 0)],
        )
        self.assertEqual(len(macro.child_macros), 0)

    async def test_wait_3(self):
        macro = parse("set(a,100).wait(10, $a).key(1)", self.context, DummyMapping, True)
        one_code = system_mapping.get("1")

        await macro.run(self.handler)
        self.assertListEqual(
            self.result,
            [(EV_KEY, one_code, 1), (EV_KEY, one_code, 0)],
        )
        self.assertEqual(len(macro.child_macros), 0)
```

Results in

```
...
test_wait_1 (tests.unit.test_macros.TestMacros.test_wait_1) ...
10:03:30.394713 231559 test.py WARNING context.py:87: Not forward_devices set
10:03:30.394864 231559 test.py WARNING context.py:90: Not source_devices set
10:03:30.396013 231559 test.py DEBUG parse.py:456: parsing macro wait(10).key(1)
10:03:30.396200 231559 test.py DEBUG parse.py:244: calls wait with 10
10:03:30.396299 231559 test.py DEBUG parse.py:244:   number 10
10:03:30.396387 231559 test.py DEBUG parse.py:244: add call to wait with [10], {}
10:03:30.396663 231559 test.py DEBUG parse.py:244: followed by key(1)
10:03:30.396757 231559 test.py DEBUG parse.py:244: calls key with 1
10:03:30.396844 231559 test.py DEBUG parse.py:244:   number 1
10:03:30.396916 231559 test.py DEBUG parse.py:244: add call to key with [1], {}
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:03:30.433010 232721 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:03:30.463682 231559 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:03:30.466676 231559 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testpxzrmnf7/.config/input-remapper-2/config.json"
10:03:30.467212 232722 test.py DEBUG shared_dict.py:63: SharedDict process started
10:03:30.468628 231559 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testpxzrmnf7/.config/input-remapper-2/config.json
10:03:30.468770 231559 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:03:30.471164 231559 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testpxzrmnf7/.config/input-remapper-2/xmodmap.json"
10:03:30.471715 231559 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testpxzrmnf7/.config/input-remapper-2/xmodmap.json"
10:03:30.477969 232722 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:03:30.478379 231559 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:03:30.478506 231559 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:03:30.478668 231559 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:03:30.478784 231559 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:03:30.478878 231559 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
test_wait_2 (tests.unit.test_macros.TestMacros.test_wait_2) ...
10:03:30.480907 231559 test.py WARNING context.py:87: Not forward_devices set
10:03:30.480997 231559 test.py WARNING context.py:90: Not source_devices set
10:03:30.481912 231559 test.py DEBUG parse.py:456: parsing macro wait(10,100).key(1)
10:03:30.482095 231559 test.py DEBUG parse.py:244: calls wait with 10,100
10:03:30.482198 231559 test.py DEBUG parse.py:244:   number 10
10:03:30.482279 231559 test.py DEBUG parse.py:244:   number 100
10:03:30.482357 231559 test.py DEBUG parse.py:244: add call to wait with [10, 100], {}
10:03:30.482627 231559 test.py DEBUG parse.py:244: followed by key(1)
10:03:30.482720 231559 test.py DEBUG parse.py:244: calls key with 1
10:03:30.482807 231559 test.py DEBUG parse.py:244:   number 1
10:03:30.482879 231559 test.py DEBUG parse.py:244: add call to key with [1], {}
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:03:30.555538 232722 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:03:30.586604 231559 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:03:30.589645 231559 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testpxzrmnf7/.config/input-remapper-2/config.json"
10:03:30.590382 232734 test.py DEBUG shared_dict.py:63: SharedDict process started
10:03:30.591841 231559 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testpxzrmnf7/.config/input-remapper-2/config.json
10:03:30.591980 231559 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:03:30.594448 231559 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testpxzrmnf7/.config/input-remapper-2/xmodmap.json"
10:03:30.594933 231559 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testpxzrmnf7/.config/input-remapper-2/xmodmap.json"
10:03:30.601400 232734 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:03:30.601777 231559 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:03:30.601886 231559 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:03:30.602029 231559 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:03:30.602142 231559 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:03:30.602242 231559 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
test_wait_3 (tests.unit.test_macros.TestMacros.test_wait_3) ...
10:03:30.604710 231559 test.py WARNING context.py:87: Not forward_devices set
10:03:30.604806 231559 test.py WARNING context.py:90: Not source_devices set
10:03:30.605807 231559 test.py DEBUG parse.py:456: parsing macro set(a,100).wait(10, $a).key(1)
10:03:30.606003 231559 test.py DEBUG parse.py:244: calls set with a,100
10:03:30.606122 231559 test.py DEBUG parse.py:244:   string a
10:03:30.606214 231559 test.py DEBUG parse.py:244:   number 100
10:03:30.606291 231559 test.py DEBUG parse.py:244: add call to set with ['a', 100], {}
10:03:30.606582 231559 test.py DEBUG parse.py:244: followed by wait(10,$a).key(1)
10:03:30.606683 231559 test.py DEBUG parse.py:244: calls wait with 10,$a
10:03:30.606775 231559 test.py DEBUG parse.py:244:   number 10
10:03:30.606869 231559 test.py DEBUG parse.py:244: add call to wait with [10, <Variable "a" at 0x7131dbda2600>], {}
10:03:30.607067 231559 test.py DEBUG parse.py:244: followed by key(1)
10:03:30.607161 231559 test.py DEBUG parse.py:244: calls key with 1
10:03:30.607252 231559 test.py DEBUG parse.py:244:   number 1
10:03:30.607333 231559 test.py DEBUG parse.py:244: add call to key with [1], {}
10:03:30.607600 231559 test.py DEBUG macro.py:569: "a" set to "100"
10:03:30.607763 232734 test.py DEBUG shared_dict.py:67: SharedDict got ('set', 'a', 100)
10:03:30.607896 232734 test.py DEBUG shared_dict.py:67: SharedDict got ('get', 'a')
10:03:30.608062 231559 test.py DEBUG macro.py:156: "<Variable "a" at 0x7131dbda2600>" is "100"
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:03:30.693082 232734 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:03:30.724958 231559 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:03:30.728065 231559 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testpxzrmnf7/.config/input-remapper-2/config.json"
10:03:30.728698 232735 test.py DEBUG shared_dict.py:63: SharedDict process started
10:03:30.730270 231559 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testpxzrmnf7/.config/input-remapper-2/config.json
10:03:30.730453 231559 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:03:30.733323 231559 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testpxzrmnf7/.config/input-remapper-2/xmodmap.json"
10:03:30.733864 231559 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testpxzrmnf7/.config/input-remapper-2/xmodmap.json"
10:03:30.740069 232735 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:03:30.740455 231559 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:03:30.740581 231559 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:03:30.740742 231559 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:03:30.740873 231559 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:03:30.740985 231559 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
```
============================================================
```
...
test_wait_1 (tests.unit.test_macros.TestMacros.test_wait_1) ...
10:05:00.041370 239412 test.py WARNING context.py:87: Not forward_devices set
10:05:00.041458 239412 test.py WARNING context.py:90: Not source_devices set
10:05:00.042370 239412 test.py DEBUG parse.py:456: parsing macro wait(10).key(1)
10:05:00.042565 239412 test.py DEBUG parse.py:244: calls wait with 10
10:05:00.042668 239412 test.py DEBUG parse.py:244:   number 10
10:05:00.042741 239412 test.py DEBUG parse.py:244: add call to wait with [10], {}
10:05:00.043011 239412 test.py DEBUG parse.py:244: followed by key(1)
10:05:00.043104 239412 test.py DEBUG parse.py:244: calls key with 1
10:05:00.043195 239412 test.py DEBUG parse.py:244:   number 1
10:05:00.043270 239412 test.py DEBUG parse.py:244: add call to key with [1], {}
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:05:00.077807 240574 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:05:00.106082 239412 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:05:00.108613 239412 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testvozjhw04/.config/input-remapper-2/config.json"
10:05:00.109386 240581 test.py DEBUG shared_dict.py:63: SharedDict process started
10:05:00.110495 239412 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testvozjhw04/.config/input-remapper-2/config.json
10:05:00.110612 239412 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:05:00.112969 239412 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testvozjhw04/.config/input-remapper-2/xmodmap.json"
10:05:00.113501 239412 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testvozjhw04/.config/input-remapper-2/xmodmap.json"
10:05:00.120213 240581 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:05:00.120594 239412 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:05:00.120693 239412 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:05:00.120826 239412 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:05:00.120936 239412 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:05:00.121030 239412 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
test_wait_2 (tests.unit.test_macros.TestMacros.test_wait_2) ...
10:05:00.123327 239412 test.py WARNING context.py:87: Not forward_devices set
10:05:00.123436 239412 test.py WARNING context.py:90: Not source_devices set
10:05:00.124439 239412 test.py DEBUG parse.py:456: parsing macro wait(10,100).key(1)
10:05:00.124632 239412 test.py DEBUG parse.py:244: calls wait with 10,100
10:05:00.124746 239412 test.py DEBUG parse.py:244:   number 10
10:05:00.124835 239412 test.py DEBUG parse.py:244:   number 100
10:05:00.124914 239412 test.py DEBUG parse.py:244: add call to wait with [10, 100], {}
10:05:00.125186 239412 test.py DEBUG parse.py:244: followed by key(1)
10:05:00.125286 239412 test.py DEBUG parse.py:244: calls key with 1
10:05:00.125436 239412 test.py DEBUG parse.py:244:   number 1
10:05:00.125560 239412 test.py DEBUG parse.py:244: add call to key with [1], {}
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:05:00.174690 240581 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:05:00.202903 239412 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:05:00.205292 239412 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testvozjhw04/.config/input-remapper-2/config.json"
10:05:00.205859 240582 test.py DEBUG shared_dict.py:63: SharedDict process started
10:05:00.207499 239412 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testvozjhw04/.config/input-remapper-2/config.json
10:05:00.207626 239412 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:05:00.209995 239412 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testvozjhw04/.config/input-remapper-2/xmodmap.json"
10:05:00.210537 239412 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testvozjhw04/.config/input-remapper-2/xmodmap.json"
10:05:00.216536 240582 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:05:00.216930 239412 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:05:00.217052 239412 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:05:00.217256 239412 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:05:00.217404 239412 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:05:00.217525 239412 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
test_wait_3 (tests.unit.test_macros.TestMacros.test_wait_3) ...
10:05:00.220417 239412 test.py WARNING context.py:87: Not forward_devices set
10:05:00.220570 239412 test.py WARNING context.py:90: Not source_devices set
10:05:00.221784 239412 test.py DEBUG parse.py:456: parsing macro set(a,100).wait(10, $a).key(1)
10:05:00.222001 239412 test.py DEBUG parse.py:244: calls set with a,100
10:05:00.222126 239412 test.py DEBUG parse.py:244:   string a
10:05:00.222217 239412 test.py DEBUG parse.py:244:   number 100
10:05:00.222295 239412 test.py DEBUG parse.py:244: add call to set with ['a', 100], {}
10:05:00.222595 239412 test.py DEBUG parse.py:244: followed by wait(10,$a).key(1)
10:05:00.222697 239412 test.py DEBUG parse.py:244: calls wait with 10,$a
10:05:00.222797 239412 test.py DEBUG parse.py:244:   number 10
10:05:00.222897 239412 test.py DEBUG parse.py:244: add call to wait with [10, <Variable "a" at 0x72e6b3661e80>], {}
10:05:00.223140 239412 test.py DEBUG parse.py:244: followed by key(1)
10:05:00.223260 239412 test.py DEBUG parse.py:244: calls key with 1
10:05:00.223369 239412 test.py DEBUG parse.py:244:   number 1
10:05:00.223466 239412 test.py DEBUG parse.py:244: add call to key with [1], {}
10:05:00.223743 239412 test.py DEBUG macro.py:569: "a" set to "100"
10:05:00.223968 240582 test.py DEBUG shared_dict.py:67: SharedDict got ('set', 'a', 100)
10:05:00.224104 240582 test.py DEBUG shared_dict.py:67: SharedDict got ('get', 'a')
10:05:00.224287 239412 test.py DEBUG macro.py:156: "<Variable "a" at 0x72e6b3661e80>" is "100"
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:05:00.273190 240582 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:05:00.303053 239412 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:05:00.305977 239412 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testvozjhw04/.config/input-remapper-2/config.json"
10:05:00.306672 240585 test.py DEBUG shared_dict.py:63: SharedDict process started
10:05:00.307924 239412 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testvozjhw04/.config/input-remapper-2/config.json
10:05:00.308055 239412 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:05:00.310652 239412 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testvozjhw04/.config/input-remapper-2/xmodmap.json"
10:05:00.311138 239412 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testvozjhw04/.config/input-remapper-2/xmodmap.json"
10:05:00.317361 240585 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:05:00.317781 239412 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:05:00.317908 239412 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:05:00.318085 239412 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:05:00.318238 239412 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:05:00.318388 239412 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
```
=====================================================
```
...
test_wait_1 (tests.unit.test_macros.TestMacros.test_wait_1) ...
10:06:38.316139 248140 test.py WARNING context.py:87: Not forward_devices set
10:06:38.316241 248140 test.py WARNING context.py:90: Not source_devices set
10:06:38.317265 248140 test.py DEBUG parse.py:456: parsing macro wait(10).key(1)
10:06:38.317530 248140 test.py DEBUG parse.py:244: calls wait with 10
10:06:38.317678 248140 test.py DEBUG parse.py:244:   number 10
10:06:38.317779 248140 test.py DEBUG parse.py:244: add call to wait with [10], {}
10:06:38.318117 248140 test.py DEBUG parse.py:244: followed by key(1)
10:06:38.318233 248140 test.py DEBUG parse.py:244: calls key with 1
10:06:38.318375 248140 test.py DEBUG parse.py:244:   number 1
10:06:38.318517 248140 test.py DEBUG parse.py:244: add call to key with [1], {}
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:06:38.354444 249316 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:06:38.385333 248140 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:06:38.388300 248140 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-test0vvotk_j/.config/input-remapper-2/config.json"
10:06:38.388878 249317 test.py DEBUG shared_dict.py:63: SharedDict process started
10:06:38.390195 248140 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-test0vvotk_j/.config/input-remapper-2/config.json
10:06:38.390332 248140 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:06:38.392813 248140 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-test0vvotk_j/.config/input-remapper-2/xmodmap.json"
10:06:38.393306 248140 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-test0vvotk_j/.config/input-remapper-2/xmodmap.json"
10:06:38.400257 249317 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:06:38.400714 248140 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:06:38.400831 248140 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:06:38.400988 248140 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:06:38.401111 248140 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:06:38.401212 248140 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
test_wait_2 (tests.unit.test_macros.TestMacros.test_wait_2) ...
10:06:38.403536 248140 test.py WARNING context.py:87: Not forward_devices set
10:06:38.403630 248140 test.py WARNING context.py:90: Not source_devices set
10:06:38.405037 248140 test.py DEBUG parse.py:456: parsing macro wait(10,100).key(1)
10:06:38.405261 248140 test.py DEBUG parse.py:244: calls wait with 10,100
10:06:38.405405 248140 test.py DEBUG parse.py:244:   number 10
10:06:38.405518 248140 test.py DEBUG parse.py:244:   number 100
10:06:38.405612 248140 test.py DEBUG parse.py:244: add call to wait with [10, 100], {}
10:06:38.405940 248140 test.py DEBUG parse.py:244: followed by key(1)
10:06:38.406061 248140 test.py DEBUG parse.py:244: calls key with 1
10:06:38.406178 248140 test.py DEBUG parse.py:244:   number 1
10:06:38.406274 248140 test.py DEBUG parse.py:244: add call to key with [1], {}
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:06:38.494163 249317 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:06:38.523665 248140 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:06:38.526675 248140 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-test0vvotk_j/.config/input-remapper-2/config.json"
10:06:38.527220 249318 test.py DEBUG shared_dict.py:63: SharedDict process started
10:06:38.529249 248140 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-test0vvotk_j/.config/input-remapper-2/config.json
10:06:38.529442 248140 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:06:38.532201 248140 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-test0vvotk_j/.config/input-remapper-2/xmodmap.json"
10:06:38.532755 248140 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-test0vvotk_j/.config/input-remapper-2/xmodmap.json"
10:06:38.539282 249318 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:06:38.539725 248140 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:06:38.539839 248140 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:06:38.539985 248140 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:06:38.540099 248140 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:06:38.540198 248140 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
test_wait_3 (tests.unit.test_macros.TestMacros.test_wait_3) ...
10:06:38.542343 248140 test.py WARNING context.py:87: Not forward_devices set
10:06:38.542428 248140 test.py WARNING context.py:90: Not source_devices set
10:06:38.543347 248140 test.py DEBUG parse.py:456: parsing macro set(a,100).wait(10, $a).key(1)
10:06:38.543540 248140 test.py DEBUG parse.py:244: calls set with a,100
10:06:38.543651 248140 test.py DEBUG parse.py:244:   string a
10:06:38.543735 248140 test.py DEBUG parse.py:244:   number 100
10:06:38.543805 248140 test.py DEBUG parse.py:244: add call to set with ['a', 100], {}
10:06:38.544066 248140 test.py DEBUG parse.py:244: followed by wait(10,$a).key(1)
10:06:38.544164 248140 test.py DEBUG parse.py:244: calls wait with 10,$a
10:06:38.544260 248140 test.py DEBUG parse.py:244:   number 10
10:06:38.544364 248140 test.py DEBUG parse.py:244: add call to wait with [10, <Variable "a" at 0x7a72edd6e000>], {}
10:06:38.544573 248140 test.py DEBUG parse.py:244: followed by key(1)
10:06:38.544663 248140 test.py DEBUG parse.py:244: calls key with 1
10:06:38.544751 248140 test.py DEBUG parse.py:244:   number 1
10:06:38.544822 248140 test.py DEBUG parse.py:244: add call to key with [1], {}
10:06:38.545036 248140 test.py DEBUG macro.py:569: "a" set to "100"
10:06:38.545211 249318 test.py DEBUG shared_dict.py:67: SharedDict got ('set', 'a', 100)
10:06:38.545367 249318 test.py DEBUG shared_dict.py:67: SharedDict got ('get', 'a')
10:06:38.545556 248140 test.py DEBUG macro.py:156: "<Variable "a" at 0x7a72edd6e000>" is "100"
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:06:38.619485 249318 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:06:38.648701 248140 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:06:38.651677 248140 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-test0vvotk_j/.config/input-remapper-2/config.json"
10:06:38.652553 249319 test.py DEBUG shared_dict.py:63: SharedDict process started
10:06:38.653554 248140 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-test0vvotk_j/.config/input-remapper-2/config.json
10:06:38.653696 248140 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:06:38.656325 248140 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-test0vvotk_j/.config/input-remapper-2/xmodmap.json"
10:06:38.656800 248140 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-test0vvotk_j/.config/input-remapper-2/xmodmap.json"
10:06:38.663067 249319 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:06:38.663485 248140 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:06:38.663595 248140 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:06:38.663736 248140 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:06:38.663847 248140 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:06:38.663940 248140 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
```
===============================
```
...
test_wait_1 (tests.unit.test_macros.TestMacros.test_wait_1) ...
10:07:13.619569 251364 test.py WARNING context.py:87: Not forward_devices set
10:07:13.619659 251364 test.py WARNING context.py:90: Not source_devices set
10:07:13.620566 251364 test.py DEBUG parse.py:456: parsing macro wait(10).key(1)
10:07:13.620743 251364 test.py DEBUG parse.py:244: calls wait with 10
10:07:13.620846 251364 test.py DEBUG parse.py:244:   number 10
10:07:13.620921 251364 test.py DEBUG parse.py:244: add call to wait with [10], {}
10:07:13.621182 251364 test.py DEBUG parse.py:244: followed by key(1)
10:07:13.621274 251364 test.py DEBUG parse.py:244: calls key with 1
10:07:13.621385 251364 test.py DEBUG parse.py:244:   number 1
10:07:13.621459 251364 test.py DEBUG parse.py:244: add call to key with [1], {}
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:07:13.656338 252478 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:07:13.685889 251364 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:07:13.688807 251364 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testrnvh5zlg/.config/input-remapper-2/config.json"
10:07:13.689411 252479 test.py DEBUG shared_dict.py:63: SharedDict process started
10:07:13.690823 251364 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testrnvh5zlg/.config/input-remapper-2/config.json
10:07:13.690962 251364 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:07:13.693376 251364 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testrnvh5zlg/.config/input-remapper-2/xmodmap.json"
10:07:13.693878 251364 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testrnvh5zlg/.config/input-remapper-2/xmodmap.json"
10:07:13.700219 252479 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:07:13.700642 251364 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:07:13.700747 251364 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:07:13.700894 251364 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:07:13.701004 251364 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:07:13.701096 251364 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
test_wait_2 (tests.unit.test_macros.TestMacros.test_wait_2) ...
10:07:13.703166 251364 test.py WARNING context.py:87: Not forward_devices set
10:07:13.703252 251364 test.py WARNING context.py:90: Not source_devices set
10:07:13.704173 251364 test.py DEBUG parse.py:456: parsing macro wait(10,100).key(1)
10:07:13.704367 251364 test.py DEBUG parse.py:244: calls wait with 10,100
10:07:13.704474 251364 test.py DEBUG parse.py:244:   number 10
10:07:13.704559 251364 test.py DEBUG parse.py:244:   number 100
10:07:13.704633 251364 test.py DEBUG parse.py:244: add call to wait with [10, 100], {}
10:07:13.704894 251364 test.py DEBUG parse.py:244: followed by key(1)
10:07:13.704990 251364 test.py DEBUG parse.py:244: calls key with 1
10:07:13.705084 251364 test.py DEBUG parse.py:244:   number 1
10:07:13.705156 251364 test.py DEBUG parse.py:244: add call to key with [1], {}
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:07:13.749502 252479 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:07:13.778926 251364 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:07:13.781848 251364 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testrnvh5zlg/.config/input-remapper-2/config.json"
10:07:13.782558 252480 test.py DEBUG shared_dict.py:63: SharedDict process started
10:07:13.784147 251364 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testrnvh5zlg/.config/input-remapper-2/config.json
10:07:13.784337 251364 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:07:13.786795 251364 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testrnvh5zlg/.config/input-remapper-2/xmodmap.json"
10:07:13.787290 251364 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testrnvh5zlg/.config/input-remapper-2/xmodmap.json"
10:07:13.793666 252480 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:07:13.794101 251364 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:07:13.794203 251364 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:07:13.794374 251364 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:07:13.794488 251364 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:07:13.794585 251364 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
test_wait_3 (tests.unit.test_macros.TestMacros.test_wait_3) ...
10:07:13.796693 251364 test.py WARNING context.py:87: Not forward_devices set
10:07:13.796779 251364 test.py WARNING context.py:90: Not source_devices set
10:07:13.797708 251364 test.py DEBUG parse.py:456: parsing macro set(a,100).wait(10, $a).key(1)
10:07:13.797931 251364 test.py DEBUG parse.py:244: calls set with a,100
10:07:13.798042 251364 test.py DEBUG parse.py:244:   string a
10:07:13.798126 251364 test.py DEBUG parse.py:244:   number 100
10:07:13.798197 251364 test.py DEBUG parse.py:244: add call to set with ['a', 100], {}
10:07:13.798473 251364 test.py DEBUG parse.py:244: followed by wait(10,$a).key(1)
10:07:13.798569 251364 test.py DEBUG parse.py:244: calls wait with 10,$a
10:07:13.798661 251364 test.py DEBUG parse.py:244:   number 10
10:07:13.798754 251364 test.py DEBUG parse.py:244: add call to wait with [10, <Variable "a" at 0x7e2da6b3ff80>], {}
10:07:13.798947 251364 test.py DEBUG parse.py:244: followed by key(1)
10:07:13.799039 251364 test.py DEBUG parse.py:244: calls key with 1
10:07:13.799136 251364 test.py DEBUG parse.py:244:   number 1
10:07:13.799212 251364 test.py DEBUG parse.py:244: add call to key with [1], {}
10:07:13.799441 251364 test.py DEBUG macro.py:569: "a" set to "100"
10:07:13.799617 252480 test.py DEBUG shared_dict.py:67: SharedDict got ('set', 'a', 100)
10:07:13.799750 252480 test.py DEBUG shared_dict.py:67: SharedDict got ('get', 'a')
10:07:13.799922 251364 test.py DEBUG macro.py:156: "<Variable "a" at 0x7e2da6b3ff80>" is "100"
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:07:13.916855 252480 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:07:13.945642 251364 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:07:13.948378 251364 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testrnvh5zlg/.config/input-remapper-2/config.json"
10:07:13.949152 252568 test.py DEBUG shared_dict.py:63: SharedDict process started
10:07:13.950312 251364 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testrnvh5zlg/.config/input-remapper-2/config.json
10:07:13.950452 251364 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:07:13.953177 251364 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testrnvh5zlg/.config/input-remapper-2/xmodmap.json"
10:07:13.953687 251364 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testrnvh5zlg/.config/input-remapper-2/xmodmap.json"
10:07:13.960029 252568 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:07:13.960428 251364 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:07:13.960531 251364 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:07:13.960679 251364 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:07:13.960793 251364 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:07:13.960890 251364 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
```

================================
```
...
test_wait_1 (tests.unit.test_macros.TestMacros.test_wait_1) ...
10:07:41.299028 253903 test.py WARNING context.py:87: Not forward_devices set
10:07:41.299121 253903 test.py WARNING context.py:90: Not source_devices set
10:07:41.300103 253903 test.py DEBUG parse.py:456: parsing macro wait(10).key(1)
10:07:41.300299 253903 test.py DEBUG parse.py:244: calls wait with 10
10:07:41.300416 253903 test.py DEBUG parse.py:244:   number 10
10:07:41.300494 253903 test.py DEBUG parse.py:244: add call to wait with [10], {}
10:07:41.300807 253903 test.py DEBUG parse.py:244: followed by key(1)
10:07:41.300906 253903 test.py DEBUG parse.py:244: calls key with 1
10:07:41.301000 253903 test.py DEBUG parse.py:244:   number 1
10:07:41.301071 253903 test.py DEBUG parse.py:244: add call to key with [1], {}
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:07:41.336474 254980 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:07:41.365866 253903 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:07:41.368994 253903 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testbekabvfv/.config/input-remapper-2/config.json"
10:07:41.369573 255069 test.py DEBUG shared_dict.py:63: SharedDict process started
10:07:41.371055 253903 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testbekabvfv/.config/input-remapper-2/config.json
10:07:41.371178 253903 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:07:41.373632 253903 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testbekabvfv/.config/input-remapper-2/xmodmap.json"
10:07:41.374114 253903 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testbekabvfv/.config/input-remapper-2/xmodmap.json"
10:07:41.380351 255069 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:07:41.380740 253903 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:07:41.380841 253903 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:07:41.380990 253903 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:07:41.381102 253903 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:07:41.381196 253903 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
test_wait_2 (tests.unit.test_macros.TestMacros.test_wait_2) ...
10:07:41.383233 253903 test.py WARNING context.py:87: Not forward_devices set
10:07:41.383324 253903 test.py WARNING context.py:90: Not source_devices set
10:07:41.384236 253903 test.py DEBUG parse.py:456: parsing macro wait(10,100).key(1)
10:07:41.384425 253903 test.py DEBUG parse.py:244: calls wait with 10,100
10:07:41.384533 253903 test.py DEBUG parse.py:244:   number 10
10:07:41.384615 253903 test.py DEBUG parse.py:244:   number 100
10:07:41.384687 253903 test.py DEBUG parse.py:244: add call to wait with [10, 100], {}
10:07:41.384954 253903 test.py DEBUG parse.py:244: followed by key(1)
10:07:41.385049 253903 test.py DEBUG parse.py:244: calls key with 1
10:07:41.385144 253903 test.py DEBUG parse.py:244:   number 1
10:07:41.385216 253903 test.py DEBUG parse.py:244: add call to key with [1], {}
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:07:41.452488 255069 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:07:41.481459 253903 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:07:41.484462 253903 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testbekabvfv/.config/input-remapper-2/config.json"
10:07:41.485037 255070 test.py DEBUG shared_dict.py:63: SharedDict process started
10:07:41.486480 253903 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testbekabvfv/.config/input-remapper-2/config.json
10:07:41.486615 253903 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:07:41.488997 253903 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testbekabvfv/.config/input-remapper-2/xmodmap.json"
10:07:41.489482 253903 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testbekabvfv/.config/input-remapper-2/xmodmap.json"
10:07:41.495623 255070 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:07:41.496007 253903 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:07:41.496109 253903 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:07:41.496252 253903 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:07:41.496373 253903 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:07:41.496477 253903 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
test_wait_3 (tests.unit.test_macros.TestMacros.test_wait_3) ...
10:07:41.498500 253903 test.py WARNING context.py:87: Not forward_devices set
10:07:41.498586 253903 test.py WARNING context.py:90: Not source_devices set
10:07:41.499528 253903 test.py DEBUG parse.py:456: parsing macro set(a,100).wait(10, $a).key(1)
10:07:41.499712 253903 test.py DEBUG parse.py:244: calls set with a,100
10:07:41.499829 253903 test.py DEBUG parse.py:244:   string a
10:07:41.499920 253903 test.py DEBUG parse.py:244:   number 100
10:07:41.499996 253903 test.py DEBUG parse.py:244: add call to set with ['a', 100], {}
10:07:41.500258 253903 test.py DEBUG parse.py:244: followed by wait(10,$a).key(1)
10:07:41.500363 253903 test.py DEBUG parse.py:244: calls wait with 10,$a
10:07:41.500456 253903 test.py DEBUG parse.py:244:   number 10
10:07:41.500551 253903 test.py DEBUG parse.py:244: add call to wait with [10, <Variable "a" at 0x7ac16e15fc50>], {}
10:07:41.500753 253903 test.py DEBUG parse.py:244: followed by key(1)
10:07:41.500848 253903 test.py DEBUG parse.py:244: calls key with 1
10:07:41.500938 253903 test.py DEBUG parse.py:244:   number 1
10:07:41.501030 253903 test.py DEBUG parse.py:244: add call to key with [1], {}
10:07:41.501247 253903 test.py DEBUG macro.py:569: "a" set to "100"
10:07:41.501416 255070 test.py DEBUG shared_dict.py:67: SharedDict got ('set', 'a', 100)
10:07:41.501572 255070 test.py DEBUG shared_dict.py:67: SharedDict got ('get', 'a')
10:07:41.501746 253903 test.py DEBUG macro.py:156: "<Variable "a" at 0x7ac16e15fc50>" is "100"
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:07:41.575011 255070 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:07:41.603523 253903 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:07:41.606444 253903 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testbekabvfv/.config/input-remapper-2/config.json"
10:07:41.607054 255071 test.py DEBUG shared_dict.py:63: SharedDict process started
10:07:41.608493 253903 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testbekabvfv/.config/input-remapper-2/config.json
10:07:41.608620 253903 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:07:41.611202 253903 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testbekabvfv/.config/input-remapper-2/xmodmap.json"
10:07:41.611866 253903 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testbekabvfv/.config/input-remapper-2/xmodmap.json"
10:07:41.618092 255071 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:07:41.618479 253903 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:07:41.618580 253903 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:07:41.618726 253903 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:07:41.618841 253903 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:07:41.618935 253903 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
```

==========================
```
...
test_wait_1 (tests.unit.test_macros.TestMacros.test_wait_1) ...
10:08:22.034974 257498 test.py WARNING context.py:87: Not forward_devices set
10:08:22.035058 257498 test.py WARNING context.py:90: Not source_devices set
10:08:22.035999 257498 test.py DEBUG parse.py:456: parsing macro wait(10).key(1)
10:08:22.036215 257498 test.py DEBUG parse.py:244: calls wait with 10
10:08:22.036343 257498 test.py DEBUG parse.py:244:   number 10
10:08:22.036424 257498 test.py DEBUG parse.py:244: add call to wait with [10], {}
10:08:22.036704 257498 test.py DEBUG parse.py:244: followed by key(1)
10:08:22.036798 257498 test.py DEBUG parse.py:244: calls key with 1
10:08:22.036888 257498 test.py DEBUG parse.py:244:   number 1
10:08:22.036963 257498 test.py DEBUG parse.py:244: add call to key with [1], {}
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:08:22.071838 258655 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:08:22.100397 257498 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:08:22.103197 257498 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testaxk1hv9d/.config/input-remapper-2/config.json"
10:08:22.103899 258656 test.py DEBUG shared_dict.py:63: SharedDict process started
10:08:22.105143 257498 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testaxk1hv9d/.config/input-remapper-2/config.json
10:08:22.105278 257498 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:08:22.107658 257498 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testaxk1hv9d/.config/input-remapper-2/xmodmap.json"
10:08:22.108125 257498 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testaxk1hv9d/.config/input-remapper-2/xmodmap.json"
10:08:22.114672 258656 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:08:22.115050 257498 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:08:22.115155 257498 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:08:22.115291 257498 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:08:22.115410 257498 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:08:22.115507 257498 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
test_wait_2 (tests.unit.test_macros.TestMacros.test_wait_2) ...
10:08:22.117585 257498 test.py WARNING context.py:87: Not forward_devices set
10:08:22.117668 257498 test.py WARNING context.py:90: Not source_devices set
10:08:22.118577 257498 test.py DEBUG parse.py:456: parsing macro wait(10,100).key(1)
10:08:22.118753 257498 test.py DEBUG parse.py:244: calls wait with 10,100
10:08:22.118857 257498 test.py DEBUG parse.py:244:   number 10
10:08:22.118939 257498 test.py DEBUG parse.py:244:   number 100
10:08:22.119008 257498 test.py DEBUG parse.py:244: add call to wait with [10, 100], {}
10:08:22.119273 257498 test.py DEBUG parse.py:244: followed by key(1)
10:08:22.119375 257498 test.py DEBUG parse.py:244: calls key with 1
10:08:22.119466 257498 test.py DEBUG parse.py:244:   number 1
10:08:22.119541 257498 test.py DEBUG parse.py:244: add call to key with [1], {}
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:08:22.154816 258656 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:08:22.183549 257498 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:08:22.186412 257498 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testaxk1hv9d/.config/input-remapper-2/config.json"
10:08:22.187209 258657 test.py DEBUG shared_dict.py:63: SharedDict process started
10:08:22.188413 257498 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testaxk1hv9d/.config/input-remapper-2/config.json
10:08:22.188551 257498 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:08:22.190894 257498 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testaxk1hv9d/.config/input-remapper-2/xmodmap.json"
10:08:22.191373 257498 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testaxk1hv9d/.config/input-remapper-2/xmodmap.json"
10:08:22.197489 258657 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:08:22.197871 257498 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:08:22.197970 257498 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:08:22.198108 257498 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:08:22.198225 257498 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:08:22.198330 257498 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
test_wait_3 (tests.unit.test_macros.TestMacros.test_wait_3) ...
10:08:22.200327 257498 test.py WARNING context.py:87: Not forward_devices set
10:08:22.200420 257498 test.py WARNING context.py:90: Not source_devices set
10:08:22.201312 257498 test.py DEBUG parse.py:456: parsing macro set(a,100).wait(10, $a).key(1)
10:08:22.201497 257498 test.py DEBUG parse.py:244: calls set with a,100
10:08:22.201607 257498 test.py DEBUG parse.py:244:   string a
10:08:22.201689 257498 test.py DEBUG parse.py:244:   number 100
10:08:22.201758 257498 test.py DEBUG parse.py:244: add call to set with ['a', 100], {}
10:08:22.202002 257498 test.py DEBUG parse.py:244: followed by wait(10,$a).key(1)
10:08:22.202094 257498 test.py DEBUG parse.py:244: calls wait with 10,$a
10:08:22.202183 257498 test.py DEBUG parse.py:244:   number 10
10:08:22.202268 257498 test.py DEBUG parse.py:244: add call to wait with [10, <Variable "a" at 0x74ae71796840>], {}
10:08:22.202475 257498 test.py DEBUG parse.py:244: followed by key(1)
10:08:22.202566 257498 test.py DEBUG parse.py:244: calls key with 1
10:08:22.202650 257498 test.py DEBUG parse.py:244:   number 1
10:08:22.202725 257498 test.py DEBUG parse.py:244: add call to key with [1], {}
10:08:22.202930 257498 test.py DEBUG macro.py:569: "a" set to "100"
10:08:22.203080 258657 test.py DEBUG shared_dict.py:67: SharedDict got ('set', 'a', 100)
10:08:22.203207 258657 test.py DEBUG shared_dict.py:67: SharedDict got ('get', 'a')
10:08:22.203382 257498 test.py DEBUG macro.py:156: "<Variable "a" at 0x74ae71796840>" is "100"
Test: macro wrote(1, 2, 1)
Test: macro wrote(1, 2, 0)
Test: Quick cleanup...
10:08:22.255237 258657 test.py DEBUG shared_dict.py:67: SharedDict got ('stop',)
10:08:22.284472 257498 test.py DEBUG shared_dict.py:57: Starting SharedDict process
10:08:22.287249 257498 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testaxk1hv9d/.config/input-remapper-2/config.json"
10:08:22.288007 258658 test.py DEBUG shared_dict.py:63: SharedDict process started
10:08:22.289161 257498 test.py INFO global_config.py:129: Saved config to /tmp/input-remapper-testaxk1hv9d/.config/input-remapper-2/config.json
10:08:22.289289 257498 test.py DEBUG system_mapping.py:128: Gathering available keycodes
10:08:22.291821 257498 test.py INFO paths.py:55: Creating file "/tmp/input-remapper-testaxk1hv9d/.config/input-remapper-2/xmodmap.json"
10:08:22.292290 257498 test.py DEBUG system_mapping.py:114: Writing "/tmp/input-remapper-testaxk1hv9d/.config/input-remapper-2/xmodmap.json"
10:08:22.298506 258658 test.py DEBUG shared_dict.py:67: SharedDict got ('ping',)
10:08:22.298878 257498 service DEBUG global_uinputs.py:132: Creating regular UInputs
10:08:22.298979 257498 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard"
10:08:22.299112 257498 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper gamepad"
10:08:22.299227 257498 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper mouse"
10:08:22.299324 257498 service DEBUG global_uinputs.py:80: creating UInput device: "input-remapper keyboard + mouse"
Test: Quick cleanup done
ok
```

If I'm reading the output correctly it seems that the macros reliably run over. `wait(10)` is already running at ~34ms. If the `wait` is looking up a variable every time it runs, as it should, the execution time adds up. I have a macro that presses a key for no more than 3000ms, but can be released early so has to loop though a `wait`ed counter. The repeated `wait`s need to add up to a lot less than 3000ms for I assume that reason.

Reply:
You'll need to do something like

```
start = time.time()
run 5 times
end = time.time()
self.assertTrue(some condition on the total time)
```

in the test. This needs to be automated.

Reply:
We're reaching the limits of my python and time so I've done what I can, which isn't overly pretty since I can't see out how to output relevant results properly from `test_macros.py` itself.

```
macro wait(10).key(1)
Runtimes min, mean, max (32.459259033203125, 32.62697060902913, 32.87649154663086)
```

```
wait(10,100).key(1)
Runtimes min, mean, max (36.56506538391113, 76.64345900217693, 121.89817428588867)
```

```
set(a,100).wait(10, $a).key(1)
Runtimes min, mean, max (37.10317611694336, 80.86951573689778, 119.4305419921875)
```

Unless I'm not understanding what you actually want, that's the time it takes the 3 macro examples to run.

Reply:
unittests are used in pipelines, like this one: https://github.com/sezanzeb/input-remapper/actions/runs/11427747182/job/31792217401 (Click on "Run tests" to see them)

They are used, so that we immediately see if something broke with our changes. In that case, github will display an error for us.

Other than that, people rarely look at the test-output. And nobody is going to look at the output of a test each time they make a commit, to verify if things still work.

This is what I mean with "automated". Unittests should automatically check if the result is as expected (time within an acceptable margin of error), and if not, raise an exception.

Here is how a test can work. It passes if 5 iterations of wait sleep between 45ms and 55ms on average.

```py
    async def test_wait_1(self):
        mapping = DummyMapping()
        mapping.macro_key_sleep_ms = 0
        macro = parse("repeat(5, wait(50))", self.context, mapping, True)

        start = time.time()
        await macro.run(self.handler)
        time_per_iteration = (time.time() - start) / 5

        self.assertLess(abs(time_per_iteration - 0.05), 0.005)
```

Reply:
I'll finish the pull request later if you don't have more time, that's fine. :)

Reply:
I've re-added the `float` inputs for the required `time` and as well as the optional `max_time` and carried that through the entire function while also making sure it didn't get flattened by the various `int`s. Also added debug output to the `add_wait` macro but it's not reflected in testing.

Using your unit testing it looks like they're all within desired +/-0.005s limits. All but a double variable retrieval are under 0.001 with`wait($a,$b)` breaking 0.002 even on my ancient work laptop.

Reply:
If you remove time time randomization from add_wait, your tests will still pass, therefore they are not able to verify if the randomization still works.

I have made some changes and a pull request to your branch: https://github.com/shawarden/input-remapper/pull/1

afterwards we can merge it. In https://github.com/shawarden/input-remapper/commit/dc98ddbf8fde46dcaae230eef02c3bb1def08efc you can see how I ended up designing the tests.